src/main/java/org/jetbrains/kotlinx/lincheck/TransformationClassLoader.java

```diff
@@ -126,8 +134,70 @@ static boolean doNotTransform(String className) {
         return cw.toByteArray();
     }
 
+    /**
+     * Returns name of class the moment before it was transformed
+     */
+    private String originalName(String className) {
+        if (className.startsWith(TRANSFORMED_PACKAGE_NAME))
+            return className.substring(TRANSFORMED_PACKAGE_NAME.length());
+        return className;
+    }
+
     @Override
     public URL getResource(String name) {
         return super.getResource(name);
     }
 }
+
+/**
+ * ClassWriter for the classes transformed by *lincheck* with a correct
+ * {@link ClassWriter#getCommonSuperClass} implementation.
+ */
+class TransformationClassWriter extends ClassWriter {
+    private final Remapper remapper;
+
+    public TransformationClassWriter(int classVersion, Remapper remapper) {
+        super(classVersion > V1_6 ? COMPUTE_FRAMES : COMPUTE_MAXS);
+        this.remapper = remapper;
+    }
+
+    /**
+     * ASM uses Class.forName for given types, however it can lead to cyclic dependencies when loading transformed classes.
+     * Thus, we call original method for not-transformed class names and then just fix it if needed.
+     */
+    @Override
+    protected String getCommonSuperClass(final String type1, final String type2) {
+        String result = super.getCommonSuperClass(originalInternalName(type1), originalInternalName(type2));
+        if (remapper != null)
+            return remapper.map(result);
+        return result;
+    }
+
+    /**
+     * Returns name of class the moment before it was transformed
+     */
+    private String originalInternalName(String className) {
+        if (className.startsWith(TRANSFORMED_PACKAGE_INTERNAL_NAME))
+            return className.substring(TRANSFORMED_PACKAGE_INTERNAL_NAME.length());
+        return className;
+    }
+}
+
+/**
+ * Visitor for retrieving information of class version needed for choosing between COMPUTE_FRAME and COMPUTE_MAXS
```

Comment:
Please explain here why do we need to make this choice

Reply:
I believe, it was an optimization for Java 6 or older classes due to the fact that these classes may not have valid stack frames. I consider removing this code